### ISSUE TITLE
Create the migration process between api-server and CRDs in Service Catalog

### ISSUE THREAD

**Issue by mszostok** [kyma-project/kyma#3779]
**Description**

We need to create a migration guide/scripts which will allow clients to do easily migraton from api-server to CRD approach.

**AC**
- migration process is well documented (3MD)
- upgrade with migration only if previous SC version has api-server (1MD)
- check SBU in case SB is deleted. We should keep them (2MD)
- block API-Server mutations operation before backup (probably by webhook functionality) (2MD)
- SPIKE: for "Helm rollback to previous release" from CRD to api-server (hook:rollback) (1MD)
- update CRD chart with migration-tool (1MD)
- provide script for which is doing more or less what our Kyma upgrade plan is doing. Designed to Service catalog upstream. (3MD)

**Comment by PK85** [comment]
Issues to be created:
- find critical points that we need to create a contracts
- more or less mock critical paths and execute upgrade with all steps

**Comment by mszostok** [comment]
remember that our solution should work both with dedicated etcd (installed by ServiceCatalog chart) and also with scenario where user used already existing etcd.

For example, we cannot have assumption that we are not triggering the deletion of the CustomResource because we will always delete the underlying storage. In some cases that could not be true and in restore phase we can have some conflict.

**Comment by piotrmiskiewicz** [comment]
PoC: manual steps which performs the migration: https://github.com/piotrmiskiewicz/service-catalog/blob/to-crd-migration/contrib/hack/crd/migration.md

**Comment by piotrmiskiewicz** [comment]
Todos: 
 - make Service Catalog CRDs version chart upgradeable, AC: the upgrade must work fine `helm upgrade catalog <path to chart>`

**Comment by piotrmiskiewicz** [comment]
Using main ETCD in Minikube:
```
./bin/minikube.sh
```

copy certs and create a proper secret:
```
kubectl cp --namespace kube-system kube-apiserver-minikube:var/lib/minikube/certs/apiserver-etcd-client.crt etcd-client.crt
kubectl cp --namespace kube-system kube-apiserver-minikube:var/lib/minikube/certs/apiserver-etcd-client.key etcd-client.key
kubectl cp --namespace kube-system kube-apiserver-minikube:var/lib/minikube/certs/etcd/ca.crt etcd-client-ca.crt

kubectl create ns kyma-system

kubectl create secret generic -n kyma-system service-catalog-apiserver-cert --from-file=etcd-client.key --from-file=etcd-client.crt --from-file=etcd-client-ca.crt
```
install SC:

```
helm repo add svc-cat https://svc-catalog-charts.storage.googleapis.com
helm install svc-cat/catalog --set apiserver.storage.etcd.servers=https://127.0.0.1:2379 --set apiserver.storage.etcd.tls.enabled=true --set apiserver.storage.etcd.tls.clientCertSecretName=service-catalog-apiserver-cert --set apiserver.storage.etcd.useEmbedded=false \
    --name catalog --namespace catalog
```

create resources, then connect directly to etcd and get it, for example:
```
kubectl port-forward -n kube-system etcd-minikube 2379:2379 &
ETCDCTL_API=3 etcdctl --cacert=etcd-client-ca.crt --key=etcd-client.key --cert=etcd-client.crt get /registry/servicecatalog.k8s.io/clusterservicebrokers/helm-broker --endpoints=https://127.0.0.1:2379
```

gives an output:
```json
/registry/servicecatalog.k8s.io/clusterservicebrokers/helm-broker
{"kind":"ClusterServiceBroker","apiVersion":"servicecatalog.k8s.io/v1beta1","metadata":{"name":"helm-broker","uid":"07aa039e-869d-11e9-8d41-9242bee4c30c","generation":1,"creationTimestamp":"2019-06-04T07:47:40Z","annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"servicecatalog.k8s.io/v1beta1\",\"kind\":\"ClusterServiceBroker\",\"metadata\":{\"annotations\":{},\"name\":\"helm-broker\"},\"spec\":{\"relistRequests\":1,\"url\":\"http://helm-broker/\"}}\n"},"finalizers":["kubernetes-incubator/service-catalog"]},"spec":{"url":"http://helm-broker/","relistBehavior":"Duration","relistRequests":1},"status":{"conditions":[{"type":"Ready","status":"False","lastTransitionTime":"2019-06-04T07:47:41Z","reason":"ErrorFetchingCatalog","message":"Error fetching catalog.Error getting broker catalog: Get http://helm-broker/v2/catalog: dial tcp: lookup helm-broker on 10.96.0.10:53: no such host"}],"reconciledGeneration":0,"operationStartTime":"2019-06-04T07:47:41Z"}}
```

**Comment by mszostok** [comment]
## Make Service Catalog CRDs version chart upgradeable

Problems encountered when executing helm upgrade against the Service Catalog release with Aggregated API Server

- Port already allocated 
  ```
  UPGRADE FAILED
  Error: failed to create resource: Service "catalog-catalog-webhook" is invalid: 
  spec.ports[0].nodePort: Invalid value: 30443: provided port is already allocated
  Error: UPGRADE FAILED: failed to create resource: Service "catalog-catalog-webhook" is invalid: spec.ports[0].nodePort: Invalid value: 30443: provided port is already allocated
  ```
  this was solved by changing the webhook port from `30443` to `31443`. Commit: https://github.com/kubernetes-incubator/service-catalog/pull/2630/commits/e5d27e2556c9d3b6c363ade607f00245c5469025

- not found error 
  ```
  Error: UPGRADE FAILED: no Secret with the name "catalog-catalog-webhook-cert" found
  ```
  [This issue was already reported in Helm](https://github.com/helm/helm/issues/1193) and fixed in version 2.14 by introducing the flag `-cleanup-on-fail`.

  This problem occurs only if the service catalog release is marked as failed because of the problem from the first point.

**Comment by PK85** [comment]
Initial Raw scenario:
- **pre-upgrade hook**
  - replace the api-server deployment with the api-server that has a read-only mode
  - backup the Service Catalog resources to the persistent volume (consider backuping etcd and second with new format that will be used for restore)
- **upgrade**
  - remove the api-server
  - remove the etcd storage
  - adjust secrets, RBAC etc.
  - upgrade controller manager
  - install webhook server
- **post-upgrade**
  - scale down the controller manager to 0  
  - restore the Service Catalog resources - Spec and Status (status is important because we do not want to trigger provisioning for already process items)
  - scale up controller manager
